lib/chef/context.rb

```diff
@@ -0,0 +1,85 @@
+# freeze_string_literal: true
+#
+# Copyright:: Copyright (c) Chef Software Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+require_relative "licensing_config"
+
+class Chef
+  class Context
+    class << self
+      def test_kitchen_context?
+        return @context unless @context.nil?
+
+        @context = false
+        return @context if context_secret.nil? || context_secret.empty?
+
+        if File.exist?(signed_file_path)
+          # Read the nonce, timestamp and signature from the file
+          received_nonce, received_timestamp, received_signature = read_file_content
+          current_time = Time.now.utc.to_i
+          # Check if the nonce is within 30 seconds of the current time
+          if (current_time - received_timestamp.to_i).abs < 30
+            if expected_signature(received_nonce, received_timestamp) == received_signature
+              @context = true
+            end
+          end
+          # Delete the file after reading the content
+          File.delete(signed_file_path)
+        end
+
+        @context
+      end
+
+      def switch_to_workstation_entitlement
+        puts "Running under Test-Kitchen: Switching License to Chef Workstation entitlement"
+        ChefLicensing.configure do |config|
+          # Reset entitlement ID to the ID of Chef Workstation
+          config.chef_entitlement_id = Chef::LicensingConfig::WORKSTATION_ENTITLEMENT_ID
+        end
+      end
+
+      private
+
+      def context_secret
+        ENV.fetch("TEST_KITCHEN_CONTEXT", "")
+      end
+
+      def signed_file_path
+        "#{Dir.tmpdir}/kitchen/#{context_secret}"
+      end
+
+      def read_file_content
+        file_content = {}
+        File.open(signed_file_path, "r:bom|utf-16le:utf-8") do |file|
+          file.each_line do |line|
+            key, value = line.strip.split(":")
+            file_content[key] = value
+          end
+        end
```

Comment:
Given that there no docs that I'm aware of on how any of this licensing stuff works, can we please get a lot of comments here.

Something (maybe kitchen enterprise?) creates a nonce, which it then... signs, and ... adds to this magical file, and then sticks the name of the magical file in the `signed_file_path`?

So... what, the idea is that kitchen validates some license and then is passing all this data to Chef so it knows not to use a license?

Can we please get docs on all of this and comments in the code?

Reply:
Hey @jaymzh ,

Whenever the customer uses the license on a node, it will be counted as a usage. If they are using it with test kitchen to test the cookbooks, this will cause them to exhaust the license usage very fast since they are not really using it. 

The solution was to identify that the chef-client was invoked from test-kitchen and in that case, use the workstation's entitlement ID, so that the test-kitchen usage will not be counted for the real usage. The only problem with this idea is that if we use an env variable or a dummy file, it will be easily spoofable. 

To make it a little spoof proof, test kitchen will create a nonce and sign it with a secret key. All the details will then transferred to the new VM and infra client will look for the file and verify the signature to confirm it was created from test-kitchen.

I've added the comments in the code, please review and let me know if you have any questions.